idaes/core/solvers/petsc.py

```diff
@@ -731,19 +731,58 @@ def calculate_time_derivatives(m, time):
-                    try:
-                        if disc_eq[i].active:
-                            v.value = 0  # Make sure there is a value
-                            calculate_variable_from_constraint(v, disc_eq[i])
-                    except KeyError:
-                        pass  # discretization equation may not exist at first time
+                parent_block = var.parent_block()
+                disc_eq = getattr(parent_block, var.local_name + "_disc_eq")
+
+                deriv_dict = dict(
+                    (key, pyo.Reference(slc))
+                    for key, slc in slice_component_along_sets(var, (time,))
+                )
+                disc_dict = dict(
+                    (key, pyo.Reference(slc))
+                    for key, slc in slice_component_along_sets(disc_eq, (time,))
+                )
+
+                for key, deriv in deriv_dict.items():
+                    # state = state_dict[key]
+                    disc_eq = disc_dict[key]
+                    for t in time:
+                        if t < between.first() or t > between.last():
+                            # Outside of integration range, skip calculation
+                            continue
+                        try:
+                            # TODO This calculates the value of the derivative even
+                            # if one of the state var values is from outside the
+                            # integration range, so long as it's initialized. Is
+                            # this the desired behavior?
+                            if disc_eq[t].active and not deriv[t].fixed:
+                                old_value = deriv[t].value
+                                deriv[t].value = 0  # Make sure there is a value
+                                calculate_variable_from_constraint(deriv[t], disc_eq[t])
+                        except KeyError as err:
+                            # Discretization and continuity equations may or may not exist at the first or last time
+                            # points depending on the method. Backwards skips first, forwards skips last, central skips
+                            # both (which means the user needs to provide additional equations)
+                            if t == time.first() or t == time.last():
+                                pass
+                            else:
+                                raise err
+                        except ValueError as err:
+                            # At edges of between, it's unclear which adjacent
+                            # values of state variables have been populated.
+                            # Therefore we might get hit with value errors.
+                            if t == between.first() or t == between.last():
+                                # Reset deriv value to old value
+                                if disc_eq[t].active and not deriv[t].fixed:
+                                    deriv[t].value = old_value
```

Comment:
@lbianchi-lbl Do you know what Pylint is complaining about here? We can only get to this logic if old_value has already been assigned in the `try` block.

Reply:
From a quick look I'd guess it might be because of `old_value` is defined inside the `if` block on L764?

Reply:
Moved it outside the `if` block, still getting the pylint error. I'm guessing it's because Pylint doesn't know that accessing `deriv[t].value` is only going to raise a `KeyError` in that context, and will never raise a `ValueError`. I assume this is a case for some Pylint ignore statement.

The test failures appear to be unrelated to this PR.